test/testlib/helm_utilities.go

```diff
@@ -0,0 +1,285 @@
+	dirPath := filepath.Join(RESULT_DIR, namespace)
+	filePath := filepath.Join(dirPath, podName+fileNameSuffix+".log")
+
+	_ = os.MkdirAll(dirPath, 0700)
+
+	f, err := os.Create(filePath)
+	require.NoError(t, err)
+	defer f.Close()
+
+	writer := io.Writer(f)
+
+	reader, err := getAppLogStreamE(t, namespace, podName, podLogOptions)
+	// avoid generating test failure just because container logs are not available
+	if _, ok := err.(*ContainersNotStarted); ok {
+		t.Logf("Skipping log collection for pod %s because no container has been started", podName)
+		return ""
+	}
+	require.NoError(t, err)
+	require.NotNil(t, reader)
+	_, err = io.Copy(writer, reader)
+	require.NoError(t, err)
+
+	t.Logf("Finished reading log file %s", filePath)
+
+	return filePath
+}
+
+func getAppLogStreamE(t *testing.T, namespace string, podName string, podLogOptions *corev1.PodLogOptions) (io.ReadCloser, error) {
+	options := k8s.NewKubectlOptions("", "", namespace)
+
+	client, err := k8s.GetKubernetesClientFromOptionsE(t, options)
+	require.NoError(t, err)
+
+	if podLogOptions.Container == "" {
+		// Select first container if not specified; otherwise the GetLogs method will fail if there are sidecars
+		pod, err := client.CoreV1().Pods(options.Namespace).Get(context.TODO(), podName, metav1.GetOptions{})
+		require.NoError(t, err)
+
+		container := findXmtpContainer(pod)
+		if container == nil {
+			return nil, &ContainersNotStarted{}
+		}
+		podLogOptions.Container = container.Name
+		for _, containerStatus := range pod.Status.ContainerStatuses {
+			// if the container is in Waiting state (e.g. because
+			// the pod is in CrashLoopBackOff state), then get the
+			// logs from the previous invocation of the container
+			if containerStatus.Name == container.Name && containerStatus.State.Waiting != nil {
+				podLogOptions.Previous = true
+			}
+		}
+		if podLogOptions.Previous {
+			t.Logf("Multiple containers found in pod %s. Getting logs from previous container %s.", podName, podLogOptions.Container)
+		} else {
+			t.Logf("Multiple containers found in pod %s. Getting logs from container %s.", podName, podLogOptions.Container)
+		}
+	}
+
+	return client.CoreV1().Pods(options.Namespace).GetLogs(podName, podLogOptions).Stream(context.TODO())
+}
+
+func doesContainerHaveLogs(container *corev1.Container, containerStatuses []corev1.ContainerStatus) bool {
+	for _, status := range containerStatuses {
+		// check the status of the container; if it is in Waiting state,
+		// then check that it has a non-0 restart count; otherwise the
+		// container has no logs to retrieve
+		if status.Name == container.Name && (status.State.Waiting == nil || status.RestartCount > 0) {
+			return true
+		}
+	}
+	return false
+}
+
+func findXmtpContainer(pod *corev1.Pod) *corev1.Container {
+	// look for any container named "xmtp" that has logs
+	for _, container := range pod.Spec.Containers {
+		if (container.Name == "xmtpd") && doesContainerHaveLogs(&container, pod.Status.ContainerStatuses) {
```

Comment:
nit: I usually default to create a const for every value like `"xmtpd"` subject to change at some point.